enterprise/app/trends/heatmap.tsx

```diff
@@ -0,0 +1,422 @@
+import React from "react";
+import moment from "moment";
+
+import { stats } from "../../../proto/stats_ts_proto";
+import { ScaleBand, scaleBand } from "d3-scale";
+import { withResizeDetector } from "react-resize-detector";
+import { pinBottomLeftOffsetFromMouse, MouseCoords, Tooltip } from "../../../app/components/tooltip/tooltip";
+
+interface HeatmapProps {
+  heatmapData: stats.GetStatHeatmapResponse;
+  width: number;
+  height: number;
+  valueFormatter: (value: number) => string;
+  metricBucketName: string;
+  metricBucketFormatter: (value: number) => string;
+  selectionCallback?: (s?: HeatmapSelection) => void;
+}
+
+interface State {
+  selectionToRender?: [TooltipData, TooltipData];
+}
+
+type TooltipData = {
+  timestamp: number;
+  timestampBucketIndex: number;
+  metric: number;
+  metricBucketIndex: number;
+  value: number;
+};
+
+export type HeatmapSelection = {
+  dateRangeMicros: number[];
+  bucketRange: number[];
```

Comment:
nit: if these are pairs, maybe type as `[startInclusive: number, endExclusive: number]` (assuming the end range is exclusive, haven't grokked that part yet :sweat_smile: )

Reply:
Yes, correct, and yes, good idea.